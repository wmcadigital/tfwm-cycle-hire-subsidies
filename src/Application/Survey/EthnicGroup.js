import { useEffect } from "react";
import { useFormState, useForm } from "react-final-form";
import PropTypes from "prop-types";

import Ethnicity, { PreferNotToSayLabel } from "./Ethnicity";
import FormSection from "../../common/FormSection";
import RadioGroup from "../../common/RadioGroup";
import Question from "../../common/Question";
import RadioButton from "../../common/RadioButton";
import FieldError from "../../common/FieldError";
import ProgressIndicator from "../../common/ProgressIndicator";
import { required } from "../../common/validation";

const EthnicGroup = ({ setEthnicGroup, setGoToPage }) => {
  const stateApi = useFormState();
  const formApi = useForm();

  const formValues = stateApi.values;
  const question = "About you";

  const error = stateApi.submitFailed ? stateApi.errors?.ethnicity : null;

  useEffect(() => {
    if (formValues.ethnicity) {
      setEthnicGroup(formValues.ethnicity);
      if (stateApi?.modified?.ethnicity) {
        formApi.mutators.setFormAttribute("specificEthnicity", undefined);
        setGoToPage(null);
      }
    }
  }, [formValues.ethnicity]);

  return (
    <FormSection>
      <ProgressIndicator
        sectionPosition="Section 2 of 2"
        sectionName="LACF Survey"
      />
      <Question text={question} />
      <p>
        Q15 - How would you describe your ethnic group? Choose one option that
        best describes your ethnic group or background
      </p>
      <RadioGroup error={error}>
        <FieldError text={error} />
        {Object.keys(Ethnicity).map((ethnicity, index) => (
          <RadioButton
            key={index}
            label={Ethnicity[ethnicity].label}
            validation={required}
            value={ethnicity}
            fieldName="ethnicity"
          />
        ))}
        <p className="wmrards-m-b-sm">Or</p>
        <RadioButton
          key="preferNotToSay"
          label={PreferNotToSayLabel["preferNotToSay"]}
          validation={required}
          value="preferNotToSay"
          fieldName="ethnicity"
        />
      </RadioGroup>
    </FormSection>
  );
};

EthnicGroup.propTypes = {
  setEthnicGroup: PropTypes.func,
  setGoToPage: PropTypes.func,
};

EthnicGroup.defaultProps = {
  setEthnicGroup: () => {},
  setGoToPage: () => {},
};

export default EthnicGroup;
