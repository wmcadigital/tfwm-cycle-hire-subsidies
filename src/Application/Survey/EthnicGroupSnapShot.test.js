import { useFormState } from "react-final-form";
import { createRenderer } from "react-test-renderer/shallow";

import EthnicGroup from "./EthnicGroup";

jest.mock("react-final-form");

describe("EthnicGroup - snapshots", () => {
  it("individual is filling out form for themselves", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<EthnicGroup />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("individual is filling out form for someone else", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "no",
        firstName: "Test",
        lastName: "Testerson",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<EthnicGroup />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an error", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
      },
      submitFailed: true,
      errors: { ethnicity: "Required" },
    }));
    const renderer = createRenderer();

    renderer.render(<EthnicGroup />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
