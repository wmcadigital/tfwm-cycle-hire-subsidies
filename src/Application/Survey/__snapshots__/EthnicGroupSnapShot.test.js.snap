// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EthnicGroup - snapshots individual is filling out form for someone else 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="LACF Survey"
    sectionPosition="Section 2 of 2"
  />
  <Question
    text="About you"
  />
  <p>
    Q25 - How would you describe your ethnic group? Choose one option that best describes your ethnic group or background
  </p>
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="ethnicity"
      label="White"
      validation={[Function]}
      value="white"
    />
    <RadioButton
      fieldName="ethnicity"
      label="Mixed or Multiple ethnic groups"
      validation={[Function]}
      value="mixed"
    />
    <RadioButton
      fieldName="ethnicity"
      label="Asian or Asian British"
      validation={[Function]}
      value="asian"
    />
    <RadioButton
      fieldName="ethnicity"
      label="Black, African, Carribean or Black British"
      validation={[Function]}
      value="black"
    />
    <RadioButton
      fieldName="ethnicity"
      label="Other"
      validation={[Function]}
      value="other"
    />
    <p
      className="wmrards-m-b-sm"
    >
      Or
    </p>
    <RadioButton
      fieldName="ethnicity"
      label="Prefer not to say"
      validation={[Function]}
      value="preferNotToSay"
    />
  </RadioGroup>
</FormSection>
`;

exports[`EthnicGroup - snapshots individual is filling out form for themselves 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="LACF Survey"
    sectionPosition="Section 2 of 2"
  />
  <Question
    text="About you"
  />
  <p>
    Q25 - How would you describe your ethnic group? Choose one option that best describes your ethnic group or background
  </p>
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="ethnicity"
      label="White"
      validation={[Function]}
      value="white"
    />
    <RadioButton
      fieldName="ethnicity"
      label="Mixed or Multiple ethnic groups"
      validation={[Function]}
      value="mixed"
    />
    <RadioButton
      fieldName="ethnicity"
      label="Asian or Asian British"
      validation={[Function]}
      value="asian"
    />
    <RadioButton
      fieldName="ethnicity"
      label="Black, African, Carribean or Black British"
      validation={[Function]}
      value="black"
    />
    <RadioButton
      fieldName="ethnicity"
      label="Other"
      validation={[Function]}
      value="other"
    />
    <p
      className="wmrards-m-b-sm"
    >
      Or
    </p>
    <RadioButton
      fieldName="ethnicity"
      label="Prefer not to say"
      validation={[Function]}
      value="preferNotToSay"
    />
  </RadioGroup>
</FormSection>
`;

exports[`EthnicGroup - snapshots renders with an error 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="LACF Survey"
    sectionPosition="Section 2 of 2"
  />
  <Question
    text="About you"
  />
  <p>
    Q25 - How would you describe your ethnic group? Choose one option that best describes your ethnic group or background
  </p>
  <RadioGroup
    error="Required"
  >
    <FieldError
      text="Required"
    />
    <RadioButton
      fieldName="ethnicity"
      label="White"
      validation={[Function]}
      value="white"
    />
    <RadioButton
      fieldName="ethnicity"
      label="Mixed or Multiple ethnic groups"
      validation={[Function]}
      value="mixed"
    />
    <RadioButton
      fieldName="ethnicity"
      label="Asian or Asian British"
      validation={[Function]}
      value="asian"
    />
    <RadioButton
      fieldName="ethnicity"
      label="Black, African, Carribean or Black British"
      validation={[Function]}
      value="black"
    />
    <RadioButton
      fieldName="ethnicity"
      label="Other"
      validation={[Function]}
      value="other"
    />
    <p
      className="wmrards-m-b-sm"
    >
      Or
    </p>
    <RadioButton
      fieldName="ethnicity"
      label="Prefer not to say"
      validation={[Function]}
      value="preferNotToSay"
    />
  </RadioGroup>
</FormSection>
`;
